actsims/batchTools.py

```diff
@@ -0,0 +1,185 @@
+        self.apply_window    = apply_window
+
+    def get_signal_sim(self, psaf, save_alm=False, save_map=False):
+        assert(psaf in self.psafs)
+        freq = psaf.split('_')[-1]
+        print "loading sims for {}".format(psaf)
+        if self.signals.has_key(psaf): 
+            print "loading precomputed sim {}".format(psaf)
+            return self.signals[psaf].copy()
+        
+        alm_patch = None
+        if self.alms_patch.has_key(psaf):  
+            print "loading precomputed alm {}".format(psaf)
+            return self.signals[psaf].copy()
+            alm_patch = self.alms_patch[psaf].copy()
+        else:
+            if len(self.alms_base) == 0: self.load_alms_base()
+            alm_patch = self.alms_base[freq].copy()
+            if self.dobeam:
+                print "apply beam for alm {}".format(psaf)
+                beam_file = os.path.join(os.path.dirname(os.path.abspath(__file__)),signal_dict['beamNames'][psaf])
+                beam_data = (np.loadtxt(beam_file))[:,1]
+                for idx in range(alm_patch.shape[0]):
+                    alm_patch[idx] = hp.sphtfunc.almxfl(alm_patch[idx].copy(), beam_data)
+            else: pass
+            if save_alm: self.alms_patch[psaf] = alm_patch.copy()
+
+        signal = self.get_template()
+        curvedsky.alm2map(alm_patch, signal, spin = [0,2], verbose=True)
+        
+        if self.apply_window:
+            print 'apply window'
+            axes = [-2, -1]
+            for idx in range(signal.shape[0]):
+                kmap   = fft.fft(signal[idx], axes=axes)
+                wy, wx = enmap.calc_window(kmap.shape)
+                wind   = wy[:,None]**1 * wx[None,:]**1
+                kmap   *= wind
+
+                signal[idx] = (fft.ifft(kmap, axes=axes, normalize=True)).real
+                del kmap
+
+        if save_map: self.signals[psaf] = signal.copy()
+        return signal
+
+    def get_noise_sim(self, psaf, seed=None, save_map=False):
+        assert(psaf in self.psafs)
+        ## should we preload noise templates
+
+        if seed is None: seed = self.sim_idx
+        patch, season, array, freq = psaf.split('_')
+        print "loading noise sims for {} seed {}".format(psaf, seed) 
+
+        if seed is not None and self.noises[psaf].has_key(seed): 
+            print "loading precomputed sim for {} seed {}".format(psaf, seed)
+            return self.noises[psaf][seed].copy()
+        else:
+            noise = simTools.getActpolSim(iterationNum=seed,
+                        patch=patch, season=season, array=array, simType='noise', 
+                        noiseDiagsOnly=self.noisediag_only, cmbSet=self.set_idx)
+            if save_map:
+                if 'pa3' not in psaf:
+                    self.noises[psaf][seed] = noise[0].copy()
+                else:
+                    psaf_temp = '%s_{}' %('_'.join([patch,season,array]))
+                    self.noises[psaf_temp.format('f090')][seed] = noise[0].copy()
```

Comment:
Is this preserving the property that the foregrounds on a patch are the same for each [patch, season, array], but not the noise?

Reply:
Yes, that's one of the main features. It precomputes raw alm_cmb + alm_fg for given (set_idx, sim_idx, patch). Then it applies the beam on the fly as needed to get sims for particular (array, season, freq). Noise is mostly a direct function call to your function in simTools.py.

Reply:
gotcha, thanks!